### ISSUE TITLE
Crashlytics fatal crash reports are not immediately sent on Android 9 and newer

### ISSUE THREAD

**Issue by CompareAndSwap** [firebase/firebase-android-sdk#1565]
### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 3.6.3
  * Firebase Component: Crashlytics
  * Component version: 17.0.0 (issue also reproduces with older Crashlytics 2.10.1)

### [REQUIRED] Step 3: Describe the problem

When a fatal crash occurs, Crashlytics' [unhandled exception handler](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java#L373) writes the crash info to storage and it starts a background task to start uploading the crash report to the Crashlytics service. The unhandled exception handler actually returns to the OS _before the upload finishes_.

Before Android 9, the OS would keep the process alive until the user dismissed the "(appname) has stopped" dialog. This would typically be enough time for the Crashlytics background task to finish uploading the crash report.

In Android 9 and newer, there is no more "(appname) has stopped" dialog and the OS kills the process once the unhandled exception handler returns to the OS. There is no time for the Crashlytics background task to finish uploading the crash report.

Thus, when the user has a more modern version of Android, the crash report isn’t sent until the next time the app starts (if the user chooses to ever run it again).

#### Root cause:

Obviously the change to Android plays a big part here, but another part is this annotated code from [CrashlyticsController.java](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java#L373):

```
  synchronized void handleUncaughtException(
      @NonNull SettingsDataProvider settingsDataProvider,
      @NonNull final Thread thread,
      @NonNull final Throwable ex) {
...
    final Task<Void> recordFatalFirebaseEventTask = recordFatalFirebaseEvent(time.getTime());
    final Task<Void> handleUncaughtExceptionTask =
        backgroundWorker.submitTask(
            new Callable<Task<Void>>() {
              @Override
              public Task<Void> call() throws Exception {
...
                return settingsDataProvider
                    .getAppSettings()
                    .onSuccessTask(
                        executor,
                        new SuccessContinuation<AppSettingsData, Void>() {
                          @NonNull
                          @Override
                          public Task<Void> then(@Nullable AppSettingsData appSettingsData)
                              throws Exception {
...
                            // sendSessionReports() starts a background task to send the
                            // crash report, but the task is not waited on by the call to
                            // Utils.awaitEvenIfOnMainThread() below.
                            sendSessionReports(appSettingsData, dataCollectionToken);
                            reportingCoordinator.sendReports(
                                executor, DataTransportState.getState(appSettingsData));
                            return recordFatalFirebaseEventTask;
                          }
                        });
              }
            });

    try {
      // This waits on handleUncaughtExceptionTask and recordFatalFirebaseEventTask and the
      // anonymous SuccessContinuation above, but it does *NOT* wait on the task started by
      // sendSessionReports().
      Utils.awaitEvenIfOnMainThread(handleUncaughtExceptionTask);
    } catch (Exception e) {
      ...
    }
  }
```

If the code at the end of `handleUncaughtException()` waited on the newly created background task (created by `sendSessionReports()` ([link](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java#L1717))), then the issue would probably be solved.

#### Steps to reproduce:

1. Run `adb shell setprop log.tag.FirebaseCrashlytics VERBOSE` to enable verbose logcat logging.
2. Run an app with Firebase Crashlytics, cause an unhandled exception.

Before Android 9, chances are the logcat output will look like:

```
2020-05-16 18:03:47.085 5301-5301/? D/AndroidRuntime: Shutting down VM
2020-05-16 18:03:47.086 5301-5301/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: (app package), PID: 5301
    java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:448)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
     Caused by: ...
2020-05-16 18:03:47.086 5301-5301/? D/FirebaseCrashlytics: Crashlytics is handling uncaught exception "java.lang.RuntimeException: java.lang.reflect.InvocationTargetException" from thread main
2020-05-16 18:03:47.089 5301-5361/? D/FirebaseCrashlytics: Skipping logging Crashlytics event to Firebase, no Firebase Analytics
2020-05-16 18:03:47.101 5301-5322/? D/FirebaseCrashlytics: No log data to include with this event.
2020-05-16 18:03:47.109 5301-5306/? I/zygote: Do partial code cache collection, code=25KB, data=28KB
2020-05-16 18:03:47.111 5301-5306/? I/zygote: After code cache collection, code=25KB, data=28KB
2020-05-16 18:03:47.112 5301-5306/? I/zygote: Increasing code cache capacity to 128KB
2020-05-16 18:03:47.115 5301-5322/? D/FirebaseCrashlytics: No log data to include with this event.
2020-05-16 18:03:47.137 5301-5322/? D/FirebaseCrashlytics: Closing open sessions.
2020-05-16 18:03:47.137 5301-5322/? D/FirebaseCrashlytics: Closing session: (guid)
2020-05-16 18:03:47.137 5301-5322/? D/FirebaseCrashlytics: Collecting session parts for ID (guid)
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Session (guid) has fatal exception: true
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Session (guid) has non-fatal exceptions: false
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Collecting SessionStart data for session ID (guid)
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Collecting SessionUser data for session ID (guid)
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Collecting SessionApp data for session ID (guid)
2020-05-16 18:03:47.138 5301-5322/? D/FirebaseCrashlytics: Collecting SessionOS data for session ID (guid)
2020-05-16 18:03:47.139 5301-5322/? D/FirebaseCrashlytics: Collecting SessionDevice data for session ID (guid)
2020-05-16 18:03:47.139 5301-5322/? D/FirebaseCrashlytics: Removing session part files for ID (guid)
2020-05-16 18:03:47.144 5301-5322/? D/FirebaseCrashlytics: Could not read user ID file in (hex)
...
2020-05-16 18:03:47.160 5301-5322/? D/FirebaseCrashlytics: Opening a new session with ID (guid2)
2020-05-16 18:03:47.166 5301-5322/? D/FirebaseCrashlytics: Send via DataTransport disabled. Removing DataTransport reports.
2020-05-16 18:03:47.167 5301-5301/? D/FirebaseCrashlytics: Crashlytics completed exception processing. Invoking default exception handler.
2020-05-16 18:03:47.168 5301-5322/? D/FirebaseCrashlytics: Attempting to send crash report at time of crash...
2020-05-16 18:03:47.177 5301-5322/? D/FirebaseCrashlytics: Adding single file (guid).cls to report (guid)
2020-05-16 18:03:47.178 5301-5322/? D/FirebaseCrashlytics: Sending report to: https://reports.crashlytics.com/spi/v1/platforms/android/apps/(app package)/reports
2020-05-16 18:03:47.646 5301-5322/? D/FirebaseCrashlytics: Create report request ID: null
2020-05-16 18:03:47.646 5301-5322/? D/FirebaseCrashlytics: Result was: 202
2020-05-16 18:03:47.646 5301-5322/? I/FirebaseCrashlytics: Crashlytics Reports Endpoint upload complete: (guid)
2020-05-16 18:03:47.646 5301-5322/? D/FirebaseCrashlytics: Removing report at /data/user/0/(app package)/files/.com.google.firebase.crashlytics/fatal-sessions/(guid).cls
```

With Android 9 and newer, chances are the logcat output will look like:

```
2020-05-16 18:12:13.174 5736-5736/? D/AndroidRuntime: Shutting down VM
2020-05-16 18:12:13.175 5736-5736/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: (app package), PID: 5736
    java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:503)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: ...
2020-05-16 18:12:13.175 5736-5736/? D/FirebaseCrashlytics: Crashlytics is handling uncaught exception "java.lang.RuntimeException: java.lang.reflect.InvocationTargetException" from thread main
2020-05-16 18:12:13.178 5736-6092/? D/FirebaseCrashlytics: Skipping logging Crashlytics event to Firebase, no Firebase Analytics
2020-05-16 18:12:13.257 5736-5757/? D/FirebaseCrashlytics: No log data to include with this event.
2020-05-16 18:12:13.358 5736-5757/? D/FirebaseCrashlytics: No log data to include with this event.
2020-05-16 18:12:13.422 5736-5757/? D/FirebaseCrashlytics: Closing open sessions.
2020-05-16 18:12:13.423 5736-5757/? D/FirebaseCrashlytics: Closing session: (guid)
2020-05-16 18:12:13.423 5736-5757/? D/FirebaseCrashlytics: Collecting session parts for ID (guid)
2020-05-16 18:12:13.423 5736-5757/? D/FirebaseCrashlytics: Session (guid) has fatal exception: true
2020-05-16 18:12:13.423 5736-5757/? D/FirebaseCrashlytics: Session (guid) has non-fatal exceptions: false
2020-05-16 18:12:13.423 5736-5757/? D/FirebaseCrashlytics: Collecting SessionStart data for session ID (guid)
2020-05-16 18:12:13.427 5736-5757/? D/FirebaseCrashlytics: Collecting SessionUser data for session ID (guid)
2020-05-16 18:12:13.427 5736-5757/? D/FirebaseCrashlytics: Collecting SessionApp data for session ID (guid)
2020-05-16 18:12:13.427 5736-5757/? D/FirebaseCrashlytics: Collecting SessionOS data for session ID (guid)
2020-05-16 18:12:13.427 5736-5757/? D/FirebaseCrashlytics: Collecting SessionDevice data for session ID (guid)
2020-05-16 18:12:13.429 5736-5757/? D/FirebaseCrashlytics: Removing session part files for ID (guid)
2020-05-16 18:12:13.440 5736-5757/? D/FirebaseCrashlytics: Could not read user ID file in (hex)
...
2020-05-16 18:12:13.494 5736-5757/? D/FirebaseCrashlytics: Opening a new session with ID (guid2)
2020-05-16 18:12:13.506 5736-5757/? D/FirebaseCrashlytics: Send via DataTransport disabled. Removing DataTransport reports.
2020-05-16 18:12:13.507 5736-5736/? D/FirebaseCrashlytics: Crashlytics completed exception processing. Invoking default exception handler.
2020-05-16 18:12:13.513 5736-5757/? D/FirebaseCrashlytics: Attempting to send crash report at time of crash...
2020-05-16 18:12:13.517 5736-5736/? I/Process: Sending signal. PID: 5736 SIG: 9
```

The relevant difference is the logs for older Android show a crash report being sent, and the logs for newer Android show the process being killed before the crash report can be sent.

**Comment by mrwillis21** [comment]
Thanks for the report @CompareAndSwap! We're aware of this issue and are in the process of working on a fix. Should have it out soon. :)

**Comment by mrwillis21** [comment]
I also want to clarify as part of this, that there are situations where we *cannot* send at crash time, and therefore Crashlytics will always make a *best effort* to send immediately, but it's not 100% guaranteed.

**Comment by CompareAndSwap** [comment]
FWIW, I tried Crashlytics 17.0.1 and the issue still seemed to remain. I looked at 56fc810, but I guess that didn't help because DataTransport is disabled as evidenced by the logcat output "Send via DataTransport disabled."

I'm guessing that at some point the Firebase service is going to switch everyone over to DataTransport and then this will be fixed. Or, there will be a future change to make `sendSessionReports()` synchronous or wait for completion of background tasks.
